spec/controllers/checkout_controller_spec.rb

```diff
@@ -205,57 +208,19 @@
   end
 
   describe "#update_failed" do
+    let(:restart_checkout) { instance_double(RestartCheckout, call: true) }
+
     before do
       controller.instance_variable_set(:@order, order)
+      allow(RestartCheckout).to receive(:new) { restart_checkout }
     end
 
     it "clears the shipping address and restarts the checkout" do
       expect(controller).to receive(:clear_ship_address)
-      expect(controller).to receive(:restart_checkout)
+      expect(restart_checkout).to receive(:call)
       expect(controller).to receive(:respond_to)
-      controller.send(:update_failed)
-    end
-  end
-
-  describe "#restart_checkout" do
-    let!(:shipment_pending) { create(:shipment, order: order, state: 'pending') }
-    let!(:payment_checkout) { create(:payment, order: order, state: 'checkout') }
-    let!(:payment_failed) { create(:payment, order: order, state: 'failed') }
-
-    before do
-      order.update_attribute(:shipping_method_id, shipment_pending.shipping_method_id)
-      controller.instance_variable_set(:@order, order.reload)
-    end
 
-    context "when the order is already in the 'cart' state" do
-      it "does nothing" do
-        expect(order).to_not receive(:restart_checkout!)
-        controller.send(:restart_checkout)
-      end
-    end
-
-    context "when the order is in a subsequent state" do
-      before do
-        order.update_attribute(:state, "payment")
-      end
-
-      # NOTE: at the time of writing, it was not possible to create a shipment with a state other than
-      # 'pending' when the order has not been completed, so this is not a case that requires testing.
-      it "resets the order state, and clears incomplete shipments and payments" do
-        expect(order).to receive(:restart_checkout!).and_call_original
-        expect(order.shipping_method_id).to_not be nil
-        expect(order.shipments.count).to be 1
-        expect(order.adjustments.shipping.count).to be 1
-        expect(order.payments.count).to be 2
-        expect(order.adjustments.payment_fee.count).to be 2
```

Comment:
I think this pre execution assertion are useful in such a test with multiple changes going on. I'd keep them.

Reply:
I don't know... the fact that you need to set assertions because you are not sure what's the state the app is in feels like a symptom to me.

Reply:
ok. it's up to you 👍 . for me it's more about readability of the test. with the assertions before you see the movement. for example, knowing that when you add shipment_pending there should be a order.adjustments.shipping is very specific.

Reply:
That's a valid point although, to me, that doesn't belong here. Let's see what others think.

Reply:
👍